src/parse.rs

```diff
@@ -180,126 +241,204 @@ impl ClosureDecorator {
+            init,
+            for_stmt
+              .test
+              .as_ref()
+              .map(|test| self.parse_expr(test))
+              .unwrap_or(undefined_expr()),
+            for_stmt
+              .update
+              .as_ref()
+              .map(|update| self.parse_expr(update))
+              .unwrap_or(undefined_expr()),
+          ],
+        );
+
+        // exit the for-loop's initializer scope
+        self.vm.exit();
+
+        node
+      }
       // for (const left in right)
-      Stmt::ForIn(for_in) => self.new_node(
-        Node::ForInStmt,
-        vec![
-          match &for_in.left {
-            VarDeclOrPat::VarDecl(var_decl) => self.parse_var_declarator(var_decl.decls.first().unwrap()),
-            VarDeclOrPat::Pat(pat) => self.parse_pat(pat),
-          },
-          self.parse_expr(&for_in.right),
-          self.parse_stmt(for_in.body.as_ref()),
-        ],
-      ),
+      Stmt::ForIn(for_in) => {
+        // create a unique scope for the contents for the variables declared in the for-in-loop's initializer
+        self.vm.enter();
+
+        let var = match &for_in.left {
+          VarDeclOrPat::VarDecl(var_decl) => {
+            // bind the for-in's variable declaration
+            // for (const k in ..)
+            //            ^
+            self.vm.bind_var_decl(var_decl);
+
+            self.parse_var_declarator(var_decl.decls.first().unwrap())
+          }
+          VarDeclOrPat::Pat(pat) => self.parse_pat(pat),
+        };
+
+        let node = new_node(
+          Node::ForInStmt,
+          vec![
+            var,
+            self.parse_expr(&for_in.right),
+            self.parse_stmt(for_in.body.as_ref()),
+          ],
+        );
+
+        // exit the for-in-loop's initializer scope
+        self.vm.exit();
+
+        node
+      }
       // for (const left of right)
-      Stmt::ForOf(for_of) => self.new_node(
-        Node::ForOfStmt,
-        vec![
-          match &for_of.left {
-            VarDeclOrPat::VarDecl(var_decl) => self.parse_var_declarator(var_decl.decls.first().unwrap()),
-            VarDeclOrPat::Pat(pat) => self.parse_pat(pat),
-          },
-          self.parse_expr(&for_of.right),
-          self.parse_stmt(for_of.body.as_ref()),
-          if for_of.await_token.is_some() { true_expr() } else { false_expr() }
-        ],
-      ),
-      Stmt::If(if_stmt) => self.new_node(
+      Stmt::ForOf(for_of) => {
+        // create a unique scope for the contents for the variables declared in the for-of-loop's initializer
+        self.vm.enter();
+
+        let var = match &for_of.left {
+          VarDeclOrPat::VarDecl(var_decl) => {
+            // bind the for-of's variable declaration
+            // for (const item of ..)
+            //             ^
+            self.vm.bind_var_decl(var_decl);
+
+            self.parse_var_declarator(var_decl.decls.first().unwrap())
+          }
+          VarDeclOrPat::Pat(pat) => self.parse_pat(pat),
```

Comment:
Patterns in ForIn and ForOf seem to be missing bind logic.

Reply:
New names are only introduced by `var`, `const` or `let`